src/vs/workbench/contrib/terminal/browser/terminalBaseContextualActions.ts

```diff
@@ -0,0 +1,114 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IAction } from 'vs/base/common/actions';
+import { localize } from 'vs/nls';
+import { IOpenerService } from 'vs/platform/opener/common/opener';
+import { ContextualMatchResult, ITerminalContextualActionOptions, ITerminalInstance } from 'vs/workbench/contrib/terminal/browser/terminal';
+import { ITerminalCommand, ITerminalProcessManager } from 'vs/workbench/contrib/terminal/common/terminal';
+
+
+export const GitSimilarOutputRegex = /most similar command is\s+([^\s]{3,})\s*/;
+export const FreePortOutputRegex = /address already in use\s*(\d\.\s*){3}(\d\s*):((\d\s*){3}\d)\s*/;
+export const GitPushOutputRegex = /.*git push --set-upstream origin (.*)\s.*/;
+export const GitCreatePrOutputRegex = /.*Create\s+a\s+pull\s+request\s+for\s+\'(.+)\'\s+on\s+GitHub\s+by\s+visiting\s*:\s+remote:\s+(https:.+pull.+)\s*/;
+
+export const GitCommandLineRegex = /git.*/;
+export const GitPushCommandLineRegex = /git\s+push/;
+export const AnyCommandLineRegex = /.{4,}/;
+
+export function gitSimilarCommand(terminalInstance: Partial<ITerminalInstance>): ITerminalContextualActionOptions {
+	return {
+		commandLineMatcher: GitCommandLineRegex,
+		outputMatcher: { lineMatcher: GitSimilarOutputRegex, anchor: 'bottom' },
+		actionName: (matchResult: ContextualMatchResult) => matchResult.outputMatch ? `Run git ${matchResult.outputMatch[1]}` : ``,
+		exitCode: 1,
+		getActions: (matchResult: ContextualMatchResult, command: ITerminalCommand) => {
+			const actions: IAction[] = [];
+			const fixedCommand = matchResult?.outputMatch?.[1];
+			if (!fixedCommand) {
+				return;
+			}
+			const label = localize("terminal.fixGitCommand", "Run git {0}", fixedCommand);
+			actions.push({
+				class: undefined, tooltip: label, id: 'terminal.fixGitCommand', label, enabled: true,
+				run: () => {
+					command.command = `git ${fixedCommand}`;
+					terminalInstance.sendText?.(command.command, true);
+				}
+			});
+			return actions;
```

Comment:
I was thinking since running a command is one of the most common things we should elevate the concept into the contextual action options. This will let us make the formatting of each quick fix consistently controlled by a single spot. Example:

```ts
getActions(matchResult: ContextualMatchResult, command: ITerminalCommand): IAction | IRunCommand;

interface IRunCommand {
  command: string;
}
```

Reply:
yeah I was thinking that too. thanks for the reminder. we will prob want more properties on `IRunCommand` so we don't have to guess though

Reply:
probably sufficient to have `IRunCommand` extend `IAction`